File: Part8/C/FunctionConcepts/CompilerErrors.c
// C Function Errors - All errors active for compilation

#include <stdio.h>
#include <stdbool.h>

// Error: Missing return type (implicit int in old C, error in modern C)
printHello() {
    printf("Hello\n");
}

// Error: Missing semicolon after function definition
void sayHi() {
    printf("Hi\n");
}

void greet() {
    printf("Greetings\n");
}

// Error: Missing parameter type in definition
float calculateTotal(size, int toppings) {
    return 15.99;
}

// Error: Return type mismatch - returning float from int function
int getValue() {
    return 3.14;
}

// Error: Missing return statement in non-void function
int getNumber() {
    int x = 5;
}

// Error: Returning value from void function
void processData() {
    return 42;
}

// Error: Using 'call' keyword before function name
void myFunction() {
    printf("Test\n");
}

void test() {
    call myFunction();
}

// Error: Missing parentheses in function call
void display() {
    printf("Display\n");
}

void run() {
    display;
}

// Error: Default parameters (not supported in C)
void orderPizza(char* size = "medium", int toppings, bool delivery = false) {
    printf("Order placed\n");
}

// Error: Named parameters in function call (not supported in C)
void setPizza(char* size, int count) {
    printf("%s %d\n", size, count);
}

void order() {
    setPizza(size = "large", count = 3);
}

// Error: Function overloading (not supported in C)
int calculate(int a, int b) {
    return a + b;
}

float calculate(float a, float b) {
    return a + b;
}

// Error: Nested function definition (not standard C)
void outerFunction() {
    void innerFunction() {
        printf("Inner\n");
    }
    innerFunction();
}

// Error: Missing semicolon after function prototype
void prototype()

// Error: Using 'begin' and 'end' keywords
void compute() begin
    int x = 5;
end

// Error: Using 'procedure' instead of 'void'
procedure doSomething() {
    printf("Something\n");
}

// Error: Semicolon after function header before opening brace
void processValue(); {
    int x = 10;
}

// Error: Missing parameter name in function definition
int addNumbers(int, int b) {
    return 5 + b;
}

// Error: Using 'none' return type
none printMessage() {
    printf("Message\n");
}

// Error: Comma instead of semicolon after function prototype
void myProto(),

// Error: Wrong number of arguments in function call
int add(int a, int b) {
    return a + b;
}

void callAdd() {
    int result = add(5);
}

// Error: Type in parameter name position for function call
float getPrice(float base, float tax) {
    return base + tax;
}

void purchase() {
    float total = getPrice(float 10.0, float 2.0);
}

// Error: Missing comma between parameters in definition
void setValues(int count float price) {
    printf("Count: %d, Price: %.2f\n", count, price);
}

// Error: Using 'return' outside of function
int globalValue = 10;
return globalValue;

int main() {
    printf("Main\n");
    return 0;
}

File: Part8/Rust/FunctionConcepts/CompilerErrors.rs
// Rust Function Errors - All errors active for compilation

// Error: Missing return type indicator (->)
fn get_number() i32 {
    42
}

// Error: Using 'void' instead of no return type
fn void print_hello() {
    println!("Hello");
}

// Error: Using 'none' for no return type
fn none say_hi() {
    println!("Hi");
}

// Error: Parameter type before name
fn add_numbers(i32 a, i32 b) -> i32 {
    a + b
}

// Error: Missing parameter type
fn calculate_total(size, toppings: i32) -> f64 {
    15.99
}

// Error: Using semicolon between parameters
fn set_pizza(size: String; count: i32) {
    println!("{} {}", size, count);
}

// Error: Missing comma between parameters
fn set_values(count: i32 price: f64) {
    println!("{} {}", count, price);
}

// Error: Using 'call' keyword
fn my_function() {
    println!("Test");
}

fn test() {
    call my_function();
}

// Error: Default parameters (not supported in standard Rust)
fn order_pizza(size: String = "medium", toppings: i32) {
    println!("Order placed");
}

// Error: Named parameters in call (not supported in Rust)
fn make_pizza(size: String, count: i32) {
    println!("{} {}", size, count);
}

fn place_order() {
    make_pizza(size = "large", count = 3);
}

// Error: Function overloading (not supported in Rust)
fn calculate(a: i32, b: i32) -> i32 {
    a + b
}

fn calculate(a: f64, b: f64) -> f64 {
    a + b
}

// Error: Missing semicolon after statement
fn process_data() {
    let x = 5
    let y = 10
}

// Error: Using 'return' incorrectly - semicolon after return value makes it void
fn get_value() -> i32 {
    return 42;
}

// Error: Missing return statement or expression
fn get_price() -> f64 {
    let price = 15.99;
}

// Error: Using 'begin' and 'end' keywords
fn compute() begin
    let x = 5;
end

// Error: Using 'function' keyword
function display() {
    println!("Display");
}

// Error: Using 'procedure' keyword
procedure run() {
    println!("Running");
}

// Error: Missing parentheses in function definition
fn show {
    println!("Show");
}

// Error: Missing parentheses in function call
fn greet() {
    println!("Greet");
}

fn call_greet() {
    greet;
}

// Error: Using 'with parameters' syntax
fn execute() {
    println!("Execute");
}

fn caller() {
    call execute with parameters();
}

// Error: Wrong number of arguments
fn add(a: i32, b: i32) -> i32 {
    a + b
}

fn call_add() {
    let result = add(5);
}

// Error: Return type mismatch
fn get_pi() -> i32 {
    3.14
}

// Error: Using 'def' keyword
def print_message() {
    println!("Message");
}

// Error: Type in function call argument
fn get_total(price: f64, tax: f64) -> f64 {
    price + tax
}

fn purchase() {
    let total = get_total(f64 10.0, f64 2.0);
}

// Error: Using := for assignment or default
fn create_order(size: String := "small") {
    println!("{}", size);
}

// Error: Missing return type with arrow but no type specified
fn calculate_sum(a: i32, b: i32) -> {
    a + b
}

// Error: Semicolon after function signature
fn do_something(); {
    println!("Something");
}

// Error: Using 'return' outside function
let global_value = 10;
return global_value;

// Error: Implicit return with semicolon (makes it unit type, not i32)
fn factorial(n: i32) -> i32 {
    if n <= 1 {
        1;
    } else {
        n * factorial(n - 1);
    }
}

// Error: Multiple return values without tuple
fn divmod(a: i32, b: i32) -> i32, i32 {
    (a / b, a % b)
}

fn main() {
    println!("Main");
}
